### ISSUE TITLE
miiocli throws Illegal instruction

### ISSUE THREAD

**Issue by grasshide** [rytilahti/python-miio#1291]
I'm on a Raspberry Pi Zero. Clean fresh install of the latest Debian.

No problems during installation:
$ pip install python-miio
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Requirement already satisfied: python-miio in /home/pi/.local/lib/python3.9/site-packages (0.5.9.2)
Requirement already satisfied: click>=7 in /home/pi/.local/lib/python3.9/site-packages (from python-miio) (8.0.3)
Requirement already satisfied: tqdm<5,>=4 in /home/pi/.local/lib/python3.9/site-packages (from python-miio) (4.62.3)
Requirement already satisfied: pytz in /home/pi/.local/lib/python3.9/site-packages (from python-miio) (2021.3)
Requirement already satisfied: android_backup<1,>=0 in /home/pi/.local/lib/python3.9/site-packages (from python-miio) (0.2.0)
Requirement already satisfied: construct<3.0.0,>=2.10.56 in /home/pi/.local/lib/python3.9/site-packages (from python-miio) (2.10.67)
Requirement already satisfied: defusedxml<1,>=0 in /home/pi/.local/lib/python3.9/site-packages (from python-miio) (0.7.1)
Requirement already satisfied: netifaces<1,>=0 in /home/pi/.local/lib/python3.9/site-packages (from python-miio) (0.11.0)
Requirement already satisfied: zeroconf<1,>=0 in /home/pi/.local/lib/python3.9/site-packages (from python-miio) (0.38.1)
Requirement already satisfied: cryptography>=35 in /home/pi/.local/lib/python3.9/site-packages (from python-miio) (36.0.1)
Requirement already satisfied: PyYAML<7,>=5 in /home/pi/.local/lib/python3.9/site-packages (from python-miio) (6.0)
Requirement already satisfied: appdirs<2,>=1 in /home/pi/.local/lib/python3.9/site-packages (from python-miio) (1.4.4)
Requirement already satisfied: croniter>=1 in /home/pi/.local/lib/python3.9/site-packages (from python-miio) (1.1.0)
Requirement already satisfied: attrs in /home/pi/.local/lib/python3.9/site-packages (from python-miio) (21.4.0)
Requirement already satisfied: python-dateutil in /home/pi/.local/lib/python3.9/site-packages (from croniter>=1->python-miio) (2.8.2)
Requirement already satisfied: cffi>=1.12 in /home/pi/.local/lib/python3.9/site-packages (from cryptography>=35->python-miio) (1.15.0)
Requirement already satisfied: pycparser in /home/pi/.local/lib/python3.9/site-packages (from cffi>=1.12->cryptography>=35->python-miio) (2.21)
Requirement already satisfied: ifaddr>=0.1.7 in /home/pi/.local/lib/python3.9/site-packages (from zeroconf<1,>=0->python-miio) (0.1.7)
Requirement already satisfied: six>=1.5 in /usr/lib/python3/dist-packages (from python-dateutil->croniter>=1->python-miio) (1.16.0)

$ python -V
Python 3.9.2

When I try to execute miiocli:
$ miiocli
Illegal instruction

I already installed pythonn-miio on an older debian version on the pi zero without problems.
Can anybody give me a hint why this won't work? Big thanks!

**Comment by rytilahti** [comment]
That illegal instruction is coming from the OS and indicates that one of the dependencies (likely cryptography) is not built for your target architecture but something else. You could try creating a virtual environment and do a re-install inside of it to see if that helps.

**Comment by grasshide** [comment]
How can this happen? I mean it is a bare naked new installed debian?
I found  the install log. None of the packages indicate ArmV7.
I fixed the PATH warning. But that is not the cause I guess.
Any ideas how to locate the "bad" package? I'm quite the noob, when it comes to pip :). 

$ pip3 install python-miio
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Collecting python-miio
  Downloading https://www.piwheels.org/simple/python-miio/python_miio-0.5.9.2-py3-none-any.whl (304 kB)
     |████████████████████████████████| 304 kB 758 kB/s
Collecting appdirs<2,>=1
  Downloading https://www.piwheels.org/simple/appdirs/appdirs-1.4.4-py2.py3-none-any.whl (9.6 kB)
Collecting tqdm<5,>=4
  Downloading https://www.piwheels.org/simple/tqdm/tqdm-4.62.3-py2.py3-none-any.whl (76 kB)
     |████████████████████████████████| 76 kB 496 kB/s
Collecting PyYAML<7,>=5
  Downloading https://www.piwheels.org/simple/pyyaml/PyYAML-6.0-cp39-cp39-linux_armv6l.whl (45 kB)
     |████████████████████████████████| 45 kB 395 kB/s
Collecting pytz
  Downloading https://www.piwheels.org/simple/pytz/pytz-2021.3-py3-none-any.whl (511 kB)
     |████████████████████████████████| 511 kB 740 kB/s
Collecting netifaces<1,>=0
  Downloading https://www.piwheels.org/simple/netifaces/netifaces-0.11.0-cp39-cp39-linux_armv6l.whl (33 kB)
Collecting click>=7
  Downloading https://www.piwheels.org/simple/click/click-8.0.3-py3-none-any.whl (97 kB)
     |████████████████████████████████| 97 kB 610 kB/s
Collecting attrs
  Downloading https://www.piwheels.org/simple/attrs/attrs-21.4.0-py2.py3-none-any.whl (60 kB)
     |████████████████████████████████| 60 kB 588 kB/s
Collecting zeroconf<1,>=0
  Downloading https://www.piwheels.org/simple/zeroconf/zeroconf-0.38.1-py3-none-any.whl (97 kB)
     |████████████████████████████████| 97 kB 600 kB/s
Collecting defusedxml<1,>=0
  Downloading https://www.piwheels.org/simple/defusedxml/defusedxml-0.7.1-py2.py3-none-any.whl (25 kB)
Collecting android_backup<1,>=0
  Downloading https://www.piwheels.org/simple/android-backup/android_backup-0.2.0-py3-none-any.whl (8.6 kB)
Collecting croniter>=1
  Downloading https://www.piwheels.org/simple/croniter/croniter-1.1.0-py2.py3-none-any.whl (25 kB)
Collecting construct<3.0.0,>=2.10.56
  Downloading https://www.piwheels.org/simple/construct/construct-2.10.67-py3-none-any.whl (59 kB)
     |████████████████████████████████| 59 kB 566 kB/s
Collecting cryptography>=35
  Downloading https://www.piwheels.org/simple/cryptography/cryptography-36.0.1-cp39-cp39-linux_armv6l.whl (2.7 MB)
     |████████████████████████████████| 2.7 MB 1.2 MB/s
Collecting python-dateutil
  Downloading https://www.piwheels.org/simple/python-dateutil/python_dateutil-2.8.2-py2.py3-none-any.whl (247 kB)
     |████████████████████████████████| 247 kB 843 kB/s
Collecting cffi>=1.12
  Downloading https://www.piwheels.org/simple/cffi/cffi-1.15.0-cp39-cp39-linux_armv6l.whl (369 kB)
     |████████████████████████████████| 369 kB 859 kB/s
Collecting pycparser
  Downloading https://www.piwheels.org/simple/pycparser/pycparser-2.21-py2.py3-none-any.whl (119 kB)
     |████████████████████████████████| 119 kB 867 kB/s
Collecting ifaddr>=0.1.7
  Downloading https://www.piwheels.org/simple/ifaddr/ifaddr-0.1.7-py2.py3-none-any.whl (10 kB)
Requirement already satisfied: six>=1.5 in /usr/lib/python3/dist-packages (from python-dateutil->croniter>=1->python-miio) (1.16.0)
Installing collected packages: pycparser, python-dateutil, ifaddr, cffi, zeroconf, tqdm, PyYAML, pytz, netifaces, defusedxml, cryptography, croniter, construct, click, attrs, appdirs, android-backup, python-miio
  WARNING: The script tqdm is installed in '/home/pi/.local/bin' which is not on PATH.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  WARNING: The scripts android-backup-pack and android-backup-unpack are installed in '/home/pi/.local/bin' which is not on PATH.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  WARNING: The scripts miio-extract-tokens, miiocli and mirobo are installed in '/home/pi/.local/bin' which is not on PATH.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
Successfully installed PyYAML-6.0 android-backup-0.2.0 appdirs-1.4.4 attrs-21.4.0 cffi-1.15.0 click-8.0.3 construct-2.10.67 croniter-1.1.0 cryptography-36.0.1 defusedxml-0.7.1 ifaddr-0.1.7 netifaces-0.11.0 pycparser-2.21 python-dateutil-2.8.2 python-miio-0.5.9.2 pytz-2021.3 tqdm-4.62.3 zeroconf-0.38.1

**Comment by grasshide** [comment]
$ python -m pip install --upgrade cryptography
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Requirement already satisfied: cryptography in /home/pi/.local/lib/python3.9/site-packages (36.0.1)
Collecting cryptography
  Using cached https://www.piwheels.org/simple/cryptography/cryptography-36.0.1-cp39-cp39-linux_armv6l.whl (2.7 MB)
  Downloading https://www.piwheels.org/simple/cryptography/cryptography-36.0.0-cp39-cp39-linux_armv6l.whl (2.7 MB)
     |████████████████████████████████| 2.7 MB 1.5 MB/s
Requirement already satisfied: cffi>=1.12 in /home/pi/.local/lib/python3.9/site-packages (from cryptography) (1.15.0)
Requirement already satisfied: pycparser in /home/pi/.local/lib/python3.9/site-packages (from cffi>=1.12->cryptography) (2.21)

**Comment by rytilahti** [comment]
I have no idea why's that happening, sorry. You could try to use strace (`strace miiocli`) and kernel logs (`dmesg`) to see if it gives some hints where it crashes.

Are other python packages working just fine?

**Comment by grasshide** [comment]
Thanks for the hints!

I ran a stacktrace and got this:
stat64("/home/pi/.local/lib/python3.9/site-packages/cryptography/hazmat/bindings/_rust/__init__.cpython-39-arm-linux-gnueabihf.so", 0xbea5fc80) = -1 ENOENT (No such file or directory)

This confirmed your initial suspect -- the crypto library.

I looked up the previous version of the cryptography library (36.0.0 vs 36.0.1) and installed it.

Everything works fine with this library! :)

Thanks for your help!!
